### PR TITLE
Barcodeless runs

### PR DESCRIPTION
**Pull request by katieemelianova**

### CODE REVIEW

File: analysis_driver/driver.py [hunk 1]
```diff
@@ -61,18 +61,22 @@ def demultiplexing_pipeline(dataset):
...
+    validation_results = sample_sheet.validate(run_info.mask)
+    if not validation_results:
         raise AnalysisDriverError('Validation failed. Check SampleSheet.csv and RunInfo.xml.')
-
-    mask = sample_sheet.generate_mask(run_info.mask)
+    if validation_results == 'barcodeless':
+        has_barcode = False
+        reader._remove_samplesheet_barcodes(input_run_folder)
+    mask = sample_sheet.generate_mask(run_info.mask, validation_results)
     app_logger.info('bcl2fastq mask: ' + mask)  # e.g: mask = 'y150n,i6,y150n'
```
Comment (tcezard) [inline analysis_driver/driver.py]: barcodeless value could be hidden in the SampleSheet class and accessed in the Crawlers

File: analysis_driver/reader/sample_sheet.py [hunk 2]
```diff
@@ -48,6 +48,22 @@ def _read_sample_sheet(sample_sheet):
     f.close()
     return None, None
```
Comment (tcezard) [inline analysis_driver/reader/sample_sheet.py:69]: Underscore is not necessary

File: analysis_driver/reader/sample_sheet.py [hunk 3]
```diff
@@ -48,6 +48,22 @@ def _read_sample_sheet(sample_sheet):
...
+        with_barcode = openfile.read()
+        data = ((with_barcode.split('[Data],\n')[1]).split('\n'))[1:]
+        barcodes = []
+        for line in data:
+            if line:
+                barcode = (line.split(',')[5])
+                barcodes.append(barcode)
+        without_barcode = with_barcode
+        for b in barcodes:
+            without_barcode = without_barcode.replace(b, '')
```
Comment (tcezard) [inline analysis_driver/reader/sample_sheet.py]: the writing should be indented back

File: analysis_driver/reader/sample_sheet.py [hunk 4]
```diff
@@ -48,6 +48,22 @@ def _read_sample_sheet(sample_sheet):
     f.close()
     return None, None
 
+def _remove_samplesheet_barcodes(data_dir):
+    with open((os.path.join(data_dir, 'SampleSheet_analysis_driver.csv'))) as openfile:
+        with_barcode = openfile.read()
+        data = ((with_barcode.split('[Data],\n')[1]).split('\n'))[1:]
+        barcodes = []
```
Comment (tcezard) [inline analysis_driver/reader/sample_sheet.py]: Try to rewrite this using csv.DictReader, csv.DictWriter [See code in _populate](https://github.com/EdinburghGenomics/Analysis-Driver/blob/master/analysis_driver/reader/sample_sheet.py#L128)

File: analysis_driver/reader/sample_sheet.py [hunk 5]
```diff
@@ -110,7 +164,11 @@ def validate(self, mask):
         """
         self.debug('Validating...')
         if not mask.barcode_len:
-            return False
```
Comment (tcezard) [inline analysis_driver/reader/sample_sheet.py:113]: barcodeless should be stored here in SampleSheet as a boolean

File: analysis_driver/report_generation/report_crawlers.py [hunk 6]
```diff
@@ -64,10 +65,13 @@ def _populate_barcode_info_from_sample_sheet(self, samplesheet):
             for sample_id_obj in proj_obj.sample_ids.values():
                 for sample in sample_id_obj.samples:
                     for lane in sample.lane.split('+'):
-                        run_element_id = '%s_%s_%s' % (self.run_id, lane, sample.barcode)
```
Comment (tcezard) [inline analysis_driver/report_generation/report_crawlers.py:67]: We should use a two element id when barcode is not present.
We should also remove the ELEMENT_BARCODE entry from the dict

**Comment by katieemelianova** [comment]
I still havent resolved the issue which is crashing the pipeline: 

barcode[ELEMENT_PC_READ_IN_LANE] = barcode[ELEMENT_NB_READS_PASS_FILTER] / reads_for_lane
KeyError: 'passing_filter_reads'

I think the reason behind this is that the barcodeless RunInfo.xml files have both 'all' and 'unknown' barcodes. Though these are filtered out in 'parse_conversion_stats' by including only 'all' barcodes, they still exist in the self.barcodes_info.

I think another step needs to be added here to filter all the 'unknown' barcodes/run element IDs from making it through this stage if the run is barcodeless.

**Comment by tcezard** [comment]
closes #127

File: analysis_driver/reader/run_info.py [hunk 7]
```diff
@@ -82,16 +83,24 @@ def add(self, read):
...
+    #            return False
+    #    return True
+
+    #def validate_barcodeless(self):
+    #    """
+    #    Check that the first and last items of self.reads are not barcodes
+    #    """
+    #    if self._is_indexed_read(self.reads[0]) or self._is_indexed_read(self.reads[-1]):
+    #        return False
+    #    return True
```
Comment (mwhamgenomics) [inline analysis_driver/reader/run_info.py]: We can remove the commented code - it doesn't do much

File: tests/assets/RunInfo2.xml [hunk 8]
```diff
@@ -0,0 +1,1060 @@
...
+        <Barcode count="5084464" sequence="unknown"/>
+      </TopUnknownBarcodes>
+    </Lane>
+    <Lane number="7">
+      <TopUnknownBarcodes>
+        <Barcode count="5103039" sequence="unknown"/>
+      </TopUnknownBarcodes>
+    </Lane>
+  </Flowcell>
+</Stats>
```
Comment (mwhamgenomics) [inline tests/assets/RunInfo2.xml]: This file is a bit big. It also looks like a ConversionStats xml, not a RunInfo

File: tests/test_crawlers.py [hunk 9]
```diff
@@ -72,33 +72,33 @@ def test_projects(self):
...
+#        with patch('analysis_driver.report_generation.report_crawlers.get_user_sample_name',
+#                   return_value='test_sample'):
+#            with patch('analysis_driver.report_generation.report_crawlers.get_sex_from_lims',
+#                       return_value='female'):
+#                with patch('analysis_driver.reader.demultiplexing_parsers.get_species_from_sample',
+#                           return_value='Homo sapiens'):
+#                    self.crawler = report_generation.SampleCrawler('test_sample', 'test_project', self.test_data)
+
+#    def test_sample(self):
+#        self.compare_jsons(self.crawler.sample, self.expected_sample)
```
Comment (mwhamgenomics) [inline tests/test_crawlers.py]: Uncomment this

**Pull request merged**